typescript/sdk/src/providers/ProviderType.ts

```diff
@@ -0,0 +1,134 @@
+import type {
+  Connection,
+  Transaction as SolTransaction,
+} from '@solana/web3.js';
+import type {
+  Contract as EV5Contract,
+  providers as EV5Providers,
+  PopulatedTransaction as EV5Transaction,
+} from 'ethers';
+// import type { Contract as Ev6Contract, Provider as Ev6Provider } from 'ethers6';
+import type {
+  GetContractReturnType,
+  PublicClient,
+  Transaction as VTransaction,
+} from 'viem';
+
+export enum ProviderType {
+  EthersV5 = 'ethers-v5',
+  // EthersV6 = 'ethers-v6', Disabled for now to simplify build tooling
+  Viem = 'viem',
+  SolanaWeb3 = 'solana-web3',
+}
+
+export type ProviderMap<Value> = Partial<Record<ProviderType, Value>>;
+
+/**
+ * Providers with discriminated union of type
+ */
+
+interface TypedProviderBase<T> {
+  type: ProviderType;
+  provider: T;
+}
+
+export interface EthersV5Provider
+  extends TypedProviderBase<EV5Providers.Provider> {
+  type: ProviderType.EthersV5;
+  provider: EV5Providers.Provider;
+}
+
+// export interface EthersV6Provider extends TypedProviderBase<Ev6Provider> {
+//   type: ProviderType.EthersV6;
+//   provider: Ev6Provider;
+// }
+
+export interface ViemProvider extends TypedProviderBase<PublicClient> {
+  type: ProviderType.Viem;
+  provider: PublicClient;
+}
+
+export interface SolanaWeb3Provider extends TypedProviderBase<Connection> {
+  type: ProviderType.SolanaWeb3;
+  provider: Connection;
+}
+
+export type TypedProvider =
+  | EthersV5Provider
+  // | EthersV6Provider
+  | ViemProvider
+  | SolanaWeb3Provider;
+
+/**
+ * Contracts with discriminated union of provider type
+ */
+
+interface TypedContractBase<T> {
+  type: ProviderType;
+  contract: T;
+}
```

Comment:
not totally clear why we need this but ill defer

Reply:
We do use the TypedProviders and now. Some of these others are my guess about what we'll need soon